pkg/core/consensus/agreement/step.go

```diff
@@ -116,20 +120,37 @@ func (s *Loop) shouldCollectNow(a message.Message, round uint64, queue *consensu
 	return true
 }
 
-func (s *Loop) createWinningBlock(hash []byte, cert *block.Certificate) (block.Block, error) {
+func (s *Loop) createWinningBlock(ctx context.Context, hash []byte, cert *block.Certificate) (block.Block, error) {
 	var cm block.Block
-	if err := s.db.View(func(t database.Transaction) error {
+	err := s.db.View(func(t database.Transaction) error {
 		var err error
 		cm, err = t.FetchCandidateMessage(hash)
 		return err
-	}); err != nil {
-		return block.Block{}, err
+	})
+
+	if err != nil {
+		cm, err = s.requestCandidate(ctx, hash)
+		if err != nil {
+			return block.Block{}, err
+		}
 	}
 
 	cm.Header.Certificate = cert
 	return cm, nil
 }
 
+func (s *Loop) requestCandidate(ctx context.Context, hash []byte) (block.Block, error) {
+	ctx, cancel := context.WithDeadline(ctx, time.Now().Add(2*time.Second))
+	// Ensure we release the resources associated to this context.
+	defer cancel()
+	cm, err := s.requestor.RequestCandidate(ctx, hash)
```

Comment:
why not returning directly the call to the RequestCandidate? If there is an error, the caller will ignore the `cm` variable

Reply:
Makes sense 👍